### DISCUSSION TITLE
Beware that any multipart field may actually be a file!

### DISCUSSION THREAD

**Discussion by agentgt** [jooby-project/jooby#3464] [General]
We were doing:

```java
ctx.form().toMultiMap();
```

It worked a good portion of the time and then would just stop working even with the same requests.

See the issue is that Undertow will appear to randomly decide which fields become files. It does not appear to always be by request length but rather some other criteria that I'm unsure of.

**So if you have a large plain text box form() may have the field in form().files() !**

@jknack I am not sure if this is a bug (hence the discussion). It makes sense but I think Undertow thinks of multipart files as just bigger form fields whereas other frameworks do some content-disposition parsing for actual files.

Consequently a real toMultiMap would look like:

```java
        @SuppressWarnings("null")
        Map<String, List<String>> formMultimap = form.toMultimap();
        /*
         * Undertow will randomly decide what fields get turned into a file. Its
         * completely unclear how it makes that decision but my guess is it is
         * based on some pool.
         * 
         * Thus we must extract the contents of multipart fields that have
         * become temp files.
         */
        var files = form.files();
        if (!files.isEmpty()) {
            formMultimap = new LinkedHashMap<String, List<String>>(formMultimap);
            for (var f : files) {
                String key = f.getName();
                // should check size to not blow up heap.
                String value = new String(f.bytes(), java.nio.charset.StandardCharsets.UTF_8);
                formMultimap.put(key, List.of(value));
            }
        }
        return formMultimap;
```

**Comment by jknack** [comment]
Fixed in #3471
